src/lib/index.js

```diff
@@ -1008,6 +1008,60 @@ lib.templateString = function(string, obj) {
     });
 };
 
+/*
+ * Substitute values from an object into a string and optionally formats them using d3-format,
+ * or fallback to associated labels.
+ *
+ * Examples:
+ *  Lib.templateString('name: %{trace}', {trace: 'asdf'}) --> 'name: asdf'
+ *  Lib.templateString('name: %{trace[0].name}', {trace: [{name: 'asdf'}]}) --> 'name: asdf'
+ *  Lib.templateString('price: %{y:$.2f}', {y: 1}) --> 'price: $1.00'
+ *
+ * @param {string}  input string containing %{...:...} template strings
+ * @param {obj}     data object containing fallback text when no formatting is specified, ex.: {yLabel: 'formattedYValue'}
+ * @param {obj}     data objects containing substitution values
+ *
+ * @return {string} templated string
+ */
+
+lib.hovertemplateString = function(string, labels) {
+    var args = arguments;
+    // Not all that useful, but cache nestedProperty instantiation
+    // just in case it speeds things up *slightly*:
+    var getterCache = {};
+
+    return string.replace(TEMPLATE_STRING_REGEX, function(match, key, format) {
+        var obj, value, i;
+        for(i = 2; i < args.length; i++) {
+            obj = args[i];
+            if(obj.hasOwnProperty(key)) {
+                value = obj[key];
+                break;
+            }
+
+            if(!SIMPLE_PROPERTY_REGEX.test(key)) {
+                // getterCache[key] = getterCache[key] || lib.nestedProperty(obj, key).get;
+                // value = getterCache[key]();
+                value = getterCache[key] || lib.nestedProperty(obj, key).get();
+                if(value) getterCache[key] = value;
+            }
+            if(value !== undefined) break;
+        }
+
+        if(value === undefined) {
+            lib.warn('Variable \'' + key + '\' in hovertemplate could not be found!');
+            value = match;
+        }
+
+        if(format) {
+            value = d3.format(format.replace(/^:/, ''))(value);
+        } else {
+            if(labels.hasOwnProperty(key + 'Label')) value = labels[key + 'Label'];
```

Comment:
Very interesting. I've never been a big fan of `hasOwnProperty` (probably because it's a long word to type out), but **look at this**:

![image](https://user-images.githubusercontent.com/6675409/48222818-89533800-e363-11e8-9141-4b4744204531.png)

![image](https://user-images.githubusercontent.com/6675409/48222827-907a4600-e363-11e8-83ab-f74ea14e3e38.png)


using http://jsbench.github.io/#1aebf699d73fb743127903c0b0a8bece

... so maybe we should start switching to `hasOwnProperty` in hot codepath :fire: